### PR TITLE
refactor: Move format and filter results to process-results

### PR DESCRIPTION
**Pull request by ipapast**
Co-authored-by: Ofek Atar <ofek.atar@snyk.io>
Co-authored-by: Francesco Mari <francesco.mari@snyk.io>
Co-authored-by: Teodora Sandu <teodora.sandu@snyk.io>

- [x] Follows [CONTRIBUTING](https://github.com/snyk/snyk/blob/master/CONTRIBUTING.md) rules

#### What does this PR do?

Small refactor to extract the “process results” step which mainly consists of extracting the severity threshold filtering from the formatResutls function.

### CODE REVIEW

File: src/cli/commands/test/iac-local-execution/index.ts [hunk 1]
```diff
@@ -124,36 +122,19 @@ export async function test(
     }
 
     const scannedFiles = await scanFiles(parsedFiles);
+    // TODO: decide if this should go into scanFiles or stay here
```
Comment (YairZ101) [inline src/cli/commands/test/iac-local-execution/index.ts:125]: Why not put it inside the `processResults` function?

File: src/cli/commands/test/iac-local-execution/process-results.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  options: IaCTestFlags,
+): Promise<{
+  filteredIssues;
+  ignoreCount;
+}> => {
+  let projectPublicIds: Record<string, string> = {};
+  let gitRemoteUrl: string | undefined;
+
+  if (options.report) {
+    ({ projectPublicIds, gitRemoteUrl } = await formatAndShareResults({
```
Comment (YairZ101) [inline src/cli/commands/test/iac-local-execution/process-results.ts:24]: Why have we moved the `formatAndShareResults` function to here?

File: src/cli/commands/test/iac-local-execution/process-results.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+  const formattedResults = formatScanResults(
+    resultsWithCustomSeverities,
+    options,
+    iacOrgSettings.meta,
+    projectPublicIds,
+    gitRemoteUrl,
+  );
+
+  return filterIgnoredIssues(policy, formattedResults);
```
Comment (YairZ101) [inline src/cli/commands/test/iac-local-execution/process-results.ts:42]: Any particular reason that the `formatScanResults` is not the last step before returning the results?

**Review by YairZ101 — COMMENTED** [review]
Left some comment below.
All of them are basically me trying to understand why we diverged from the [suggested flow](https://miro.com/app/board/uXjVOI3jBlw=/?moveToWidget=3458764520615383560&cot=10) (not saying we shouldn't but if we are going to diverge from it I would like to understand what was the reason behind it).

File: src/cli/commands/test/iac-local-execution/process-results.ts (hunk 2 above)
Comment (teodora-sandu) [reply src/cli/commands/test/iac-local-execution/process-results.ts]: I think we did this temporarily. We will modify `scanFiles` to return `ScanResults` and then send those directly to `/iac-cli-share-results` (or whatever the endpoints was called). We only had an hour and wanted to keep changes as simple as we can.

File: src/cli/commands/test/iac-local-execution/process-results.ts (hunk 3 above)
Comment (teodora-sandu) [reply src/cli/commands/test/iac-local-execution/process-results.ts]: Do you see `filterIgnoredIssues` being in the main flow? When we created the flow we included them in `processResults`

File: src/cli/commands/test/iac-local-execution/process-results/process-results.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+export const processResults = async (
+  resultsWithCustomSeverities: IacFileScanResult[],
+  orgPublicId: string,
+  iacOrgSettings: IacOrgSettings,
+  policy: Policy | undefined,
+  tags: Tag[] | undefined,
+  attributes: ProjectAttributes | undefined,
+  options: IaCTestFlags,
+): Promise<{
+  filteredIssues;
```
Comment (teodora-sandu) [inline src/cli/commands/test/iac-local-execution/process-results/process-results.ts:17]: Should add typing

File: src/cli/commands/test/iac-local-execution/index.ts (hunk 1 above)
Comment (teodora-sandu) [reply src/cli/commands/test/iac-local-execution/index.ts:125]: I'm not sure. I remember trying it and then for some reason we started discussing whether it should be part of `scanFiles` or not. In the SCM import case for example (based on the Seldon work going on), these would be applied in the "scanning" phase and then the persistence of the rules (which would happen for example when the `--report` flag is used) doesn't need to care about them

So to me it made sense that this `scanFiles` function would eventually return the `ScanResults` with the actual severities

We could think about all of this though when we refactor `scanFiles` so I don't feel too strongly about it now

File: src/cli/commands/test/iac-local-execution/process-results.ts (hunk 3 above)
Comment (YairZ101) [reply src/cli/commands/test/iac-local-execution/process-results.ts]: > Do you see `filterIgnoredIssues` being in the main flow? 

Not at all, let me rephrase my question above.
In the suggested flow we've said that `formatScanResults` will be the last step inside `processResults`, currently the last step is `formatScanResults`.
Is it because we've decided to change the order of the functions inside `processResults`? didn't have time to work on it and we plan to do it in the next session?

File: src/cli/commands/test/iac-local-execution/index.ts (hunk 1 above)
Comment (francescomari) [reply src/cli/commands/test/iac-local-execution/index.ts:125]: @YairZ101 I think there is some tension between this comment and the following about `formatAndShareResults()`. On one hand, according to the suggested flow, `applyCustomSeverities()` should be part of `processResults()`. On the other hand, `formatAndShareResults()` works on the "scan results with custom severities" and not on the plain "scan results", like the diagram suggests.

In other words, the code and the suggested flow are not compatible. Either we move `applyCustomSeverities()` to the "Scan" step, or we move `formatAndShareResults()` to the "Process Results" step. I'm personally more inclined for the second option, because sharing the results sounds like a post-scanning activity to me. What are your preferences?

File: src/cli/commands/test/iac-local-execution/process-results.ts (hunk 2 above)
Comment (francescomari) [reply src/cli/commands/test/iac-local-execution/process-results.ts]: See [my comment](https://github.com/snyk/cli/pull/3080#discussion_r842495705) above.

File: src/cli/commands/test/iac-local-execution/process-results.ts (hunk 3 above)
Comment (francescomari) [reply src/cli/commands/test/iac-local-execution/process-results.ts]: We didn't have time to dig deeper in the implementation. As far as I can see, swapping the order of `formatScanResults()` and `filterIgnoredIssues()` might not be trivial. `filterIgnoredIssues()` has been written to accept an array of `FormattedResult` while, according to the suggested flow, should instead work on an array of `IacFileScanResult`. I would better leave this topic for a future session.

File: src/cli/commands/test/iac-local-execution/process-results/process-results.ts (hunk 4 above)
Comment (francescomari) [reply src/cli/commands/test/iac-local-execution/process-results/process-results.ts]: Done at cbb92e5f.

**Comment by francescomari** [comment]
> Small question: do you all think we should be adding the share-results-formatters.spec.ts file in a process-results folder in test to match the src directory?

I'm not sure what's the policy here. If we had to follow the `src/` structure as closely as possible, shouldn't the directory be `test/jest/unit/cli/commands/test/iac-local-execution/process-results/`?.

**Comment by ipapast** [comment]
> > Small question: do you all think we should be adding the share-results-formatters.spec.ts file in a process-results folder in test to match the src directory?
> 
> I'm not sure what's the policy here. If we had to follow the `src/` structure as closely as possible, shouldn't the directory be `test/jest/unit/cli/commands/test/iac-local-execution/process-results/`?.

true. we don't currently follow any policy I think. maybe a chat for another day

File: src/cli/commands/test/iac-local-execution/index.ts (hunk 1 above)
Comment (teodora-sandu) [reply src/cli/commands/test/iac-local-execution/index.ts:125]: Do we want to leave this TODO comment here? I added it but thought we would address it as part of this refactoring. If we think it's useful to leave because we'll fix it in a future PR then that's okay

File: src/cli/commands/test/iac-local-execution/process-results/share-results-formatter.ts [hunk 5]
```diff
@@ -1,9 +1,9 @@
-import { computePaths } from './file-utils';
+import { computePaths } from '../file-utils';
```
Comment (teodora-sandu) [inline src/cli/commands/test/iac-local-execution/process-results/share-results-formatter.ts:1]: Can we also move the test file `test/jest/unit/iac/share-results-formatters.spec.ts` to `test/jest/unit/iac/process-results/share-results-formatters.spec.ts`

File: src/cli/commands/test/iac-local-execution/process-results/process-results.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
+import { filterIgnoredIssues } from '../policy';
+import { formatAndShareResults } from './share-results';
+import { formatScanResults } from '../measurable-methods';
```
Comment (teodora-sandu) [inline src/cli/commands/test/iac-local-execution/process-results/process-results.ts:3]: Can we also move the import in `measure-methods` for `formatScanResults` from `./results-formatter` to `./process-results/results-formatter`. Or was there a reason why this wasn't includes in your commit @ipapast ?

If we end up moving it, could we also move the test file `test/jest/unit/iac/results-formatter.spec.ts` to `test/jest/unit/iac/process-results/results-formatter.spec.ts` and its fixture file

File: src/cli/commands/test/iac-local-execution/process-results/process-results.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
+import { filterIgnoredIssues } from '../policy';
```
Comment (teodora-sandu) [inline src/cli/commands/test/iac-local-execution/process-results/process-results.ts:1]: Does it make sense to rename this file to `index.ts` so it's clear that it's the entrypoint for this folder?

Comment (teodora-sandu) [reply src/cli/commands/test/iac-local-execution/process-results/process-results.ts:1]: Also, why not move `src/cli/commands/test/iac-local-execution/policy.ts` to `src/cli/commands/test/iac-local-execution/process-reuslts/policy.ts` too since it's part of the "processResults` flow?

(its test file and fixture file too)

**Review by teodora-sandu — CHANGES_REQUESTED** [review]
Thanks for driving this, I've left some comments but let me know what you think of them!

File: src/cli/commands/test/iac-local-execution/process-results/share-results-formatter.ts (hunk 5 above)
Comment (teodora-sandu) [reply src/cli/commands/test/iac-local-execution/process-results/share-results-formatter.ts:1]: I saw Ilianna also asked this. Although most people don't follow this rule, it's definitely strongly recommended and it will help us in the future when looking for tests

File: src/cli/commands/test/iac-local-execution/process-results/process-results.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+import { Policy } from '../../../../../lib/policy/find-and-load-policy';
+import { ProjectAttributes, Tag } from '../../../../../lib/types';
+import {
+  FormattedResult,
+  IacFileScanResult,
+  IacOrgSettings,
+  IaCTestFlags,
+} from '../types';
+
+export async function processResults(
```
[the remark this answers is not in the document]
Comment (francescomari) [reply src/cli/commands/test/iac-local-execution/process-results/process-results.ts:13]: I don't think it's worth it. All the functions used here seem to have their own unit tests, and the functionality as a whole has a lot of acceptance tests. We should be covered.

File: src/cli/commands/test/iac-local-execution/process-results/process-results.ts (hunk 7 above)
Comment (francescomari) [reply src/cli/commands/test/iac-local-execution/process-results/process-results.ts]: Fixed both in one of my latest commits.

File: src/cli/commands/test/iac-local-execution/index.ts (hunk 1 above)
Comment (francescomari) [reply src/cli/commands/test/iac-local-execution/index.ts:112]: I agree on leaving the comment there, for the moment. The refactoring for the proposed flow is not over yet, and this will leave some direction for the future.

File: src/cli/commands/test/iac-local-execution/process-results/share-results-formatter.ts (hunk 5 above)
Comment (francescomari) [reply src/cli/commands/test/iac-local-execution/process-results/share-results-formatter.ts:1]: Fix in one of my latest commits.

**Review by teodora-sandu — APPROVED** [review]
LGTM 👍
